proposals/0000-hybrid-output.md

```diff
@@ -0,0 +1,139 @@
+
+The Vite plugin that detects `export const output` will be unit tested against all valid `export` formats outlined above, as well as invalid dynamic exports.
+
+# Drawbacks
+
+- A new output mode will increase the complexity and maintenance costs of our `build` and `adapter` implementations.
+- The `export const output` syntax is JS-like, but since it requires static analysis, it does not support the full dynamicism of JS that users might expect. Clear errors will help ease this problem.
+- Having a separate mode for `hybrid` increases cognitive overhead and the learning curve of Astro.
+- `hybrid` output should arguably be the default `output` mode (similar to Next.js), but to avoid a breaking change we're going with an opt-in adoption approach
+
+# Alternatives
+
+There are many possible user-facing APIs for exposing control over `output`. We considered many alternatives, but settled on `export const output` as the one with the most favorable tradeoffs.
+
+### A. Magical exported functions (ala Next's `getStaticPaths`)
+
+**Pros** entirely statically analyzable (export names must be static)
+**Cons** cognitive overhead
+
+### B. Config-based routes
+
+**Pros** explicit, granular, debuggable
+**Cons** duplicated logic, need to keep config in sync with routes, _we already have file-based routing_
+
+### C. Exported `config` file annotations (`export const config = { output: "server" }`)
+
+**Pros** clear, works well with file-based routing
+**Cons** "uncanny valley" problem (not a real object, must be entirely statically analyzable)
+
+### D. Introduce novel syntax
+
+**Pros** clear, easily explainable as fully static
+**Cons** Not supported for API Endpoints, costly implementation, costly ecosystem migration for syntax/highlighing changes, additional new concepts for newcomers
+
+```astro
+--- output=server
+---
+<h1>Hello world</h1>
+```
+
+### E. Introduce this feature _under_ `output: 'server'` mode
+
+**Pros** one less concept to learn, no third-party updates needed
+**Cons** less clear adoption story, introduces yet another different feature set between `'static' | 'server'`, defaults to `server` rather than `static` for all routes, (hybrid mode should be a granular switch from static => server rather than an opt-in to static) 
+
+
+# Adoption strategy
+
+- Adoption will be entirely opt-in by setting `output: 'hybrid'` in your `astro.config.mjs` file.
+- This is not a breaking change, it is new behavior
+- Third-party adapters will need to coordinate changes to support `hybrid` output, but for the most part the logic will be very similar to the existing `server` output.
+
+# Unresolved questions
+
+- Should API Endpoints default to `output = "server"`? This seems more ergonomic than manually opting every API Endpoint into `'server'` output...
```

Comment:
RSS may be the example where `output = "static"` still makes sense for the default in an API endpoint

Reply:
Could Astro be always `hybrid`? Such that the `output` option in the `astro.config` file sets the default output for each file and the default mode can be overridden by exporting the `output` from the frontmatter. 🤔

- This would be the most flexible approach imho and also would be quite logical and reducing the learning curve. 

- I am totally unclear how the output would look like and how we would handle adapters as we always need an ssr-adapter to do anything on the server. E.g. we would always export a directory with `client` and `server` folders, which is weird if there is no ssr happening.